src/main/scala/tsiHost/TsiHostWidget.scala

```diff
@@ -0,0 +1,171 @@
+package testchipip
+
+import chisel3._
+import chisel3.util._
+
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.subsystem.{BaseSubsystem}
+import freechips.rocketchip.regmapper.{HasRegMap, RegField}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util._
+
+import SerialAdapter._
+
+/**
+ * TSI Host parameter class
+ *
+ * @param mmioRegWidth size of the MMIO data being sent back and forth in bits
+ * @param serialIfWidth size of the serialIO *out* of the widget
+ * @param txQueueEntries size of the queue for sending TSI requests
+ * @param rxQueueEntries size of the queue for recieving TSI responses
+ * @param baseAddress start address of the MMIO registers
+ * @param tlAddressSet set of addresses that you can access on the target side
+ * @param tlRegionType is cache of TL transcations happening on this side?
+ * @param tlNumTransactions is the number of inflight transactions that can happen (also determines the id associated with that transaction)
+ * @param tlLineBytes you can transfer up to a line at a time (this determines the amount of beats in a TL burst)
+ * @param bypassMMIO disconnect the mmio and have a bypass port that connects directly to the SerialAdapter
+ */
+case class TSIHostParams(
+  mmioRegWidth: Int = 32,
+  serialIfWidth: Int = 32,
+  txQueueEntries: Int = 16,
+  rxQueueEntries: Int = 16,
+  baseAddress: BigInt = BigInt(0x10017000),
+  tlAddressSet: AddressSet = AddressSet(0, BigInt("FFFFFFFF", 16)),
+  tlRegionType: RegionType.T = RegionType.UNCACHED,
+  tlNumTransactions: Int = 1,
+  tlLineBytes: Int = 64,
+  bypassMMIO: Boolean = false
+)
+
+/**
+ * Offsets for the mmio communication queues (base + offset to get the proper address)
+ */
+object TSIHostWidgetCtrlRegs {
+  val txQueueOffset = 0x00
+  val rxQueueOffset = 0x04
+}
+
+/**
+ * I/O to the outside world. This is the stream data out of the TSIHostWidget.
+ *
+ * @param w width in bits of connection to outside world
+ */
+class TSIHostWidgetIO(val w: Int)(implicit val p: Parameters) extends Bundle {
+  val serial = new SerialIO(w)
+  val debug: SerialIO = if(p(PeripheryTSIHostKey).bypassMMIO) { new SerialIO(w) } else { null } // unconnected in normal operation
```

Comment:
This might not be needed due to comment above, but some feedback here:

`null` is a big scala no-no. Use an `Option`:
`val debug: Option[SerialIO] = p(PeripheryTSIHostKey).option(new SerialIO(w))`

You'll need to call `.get` on places where you use `debug`